Add Session Pro as a concept to libsession

Doy-lee opened the pull request:
Proofs that are signed by the Session Pro backend are now present in the library as a concept. It's a class that contains the fields that the Session Pro Backend will produce that is encapsulated in `ProProof`. Proofs can be verified and validated and embedded into messages. This PR goes ontop of: https://github.com/session-foundation/libsession-util/pull/61

The bulk of this code is fulfilling the requirement


> Clients need a way to append Pro proofs and features to messages which use Pro features.
> 
> Libsession should have a function which takes
>
> Message plaintext
> Message type
> Destination
> Pro Features used
>
> Clients will call this function when they want to send any message, and libSession will return the encrypted message (with protobuf wrappers) which the client can then use their networking code or libSession to send to the destination.
> 
>When a client specifies to libsession that a message will be using Pro features the resulting encrypted message libsession produces will contain the Pro proof and a bitmask for Pro features used in the message.

This main functionality is in a new file `session_protocol.cpp` whose intent is to be high level wrappers over the lower level cryptography in `session_encrypt.cpp`. The main work horse functions are

```cpp

PRO_FEATURES get_pro_features_for_msg(size_t msg_size, PRO_FEATURES flags);

EncryptedForDestination encrypt_for_destination(
        std::span<const uint8_t> plaintext,
        std::span<const uint8_t> ed25519_privkey,
        const Destination& dest,
        config::Namespace space);

DecryptedEnvelope decrypt_envelope(
        const DecryptEnvelopeKey& keys,
        std::span<const uint8_t> envelope_payload,
        std::chrono::sys_seconds unix_ts,
        const array_uc32& pro_backend_pubkey);
```

Encrypting and decrypt the envelope was implemented by referencing the following implementations across the platforms:

iOS:
- https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L422-L500
- https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageReceiver.swift#L91-L174

Desktop: https://github.com/session-foundation/session-desktop/blob/4f51c28cc0ba999df0b46eb09d83bdfa9b372ab5/ts/session/sending/MessageQueue.ts

Android: https://github.com/session-foundation/session-android/blob/e4c16ae8aa8cae2f75aba565b890082563b07bd2/app/src/main/java/org/session/libsession/messaging/sending_receiving/MessageSender.kt#L398

Updating protobufs to include the new pro metadata, the main outline of those changes: 

```diff
message Envelope {
+   optional bytes  proSig = 11;
}

message Content {
+   optional ProMessage proMessage = 16;
}

+ message ProProof {
+   optional uint32 version           = 1;
+   optional bytes  genIndexHash      = 2;
+   optional bytes  rotatingPublicKey = 3;
+   optional uint64 expiryUnixTs      = 4;
+   optional bytes  sig               = 5;
+ }
+ 
+ message ProMessage {
+   optional ProProof proof    = 1;
+   optional uint64   features = 2;
+ }
```

--

The unit tests have examples of how to use the API. The idea here is that the platforms will fill in the pro metadata if they have some into the protobufs. You can then hand over the serialised protobuf `plaintext` for encryption by configuring the `Destination` as to where the message is intended to go (1o1, legacy groups or groups v2 essentially) with the necessary keys and call `encrypt_for_destination`. 

You then call `decrypt_envelope` when you receive a payload from the wire to get the plaintext content as well as the pro metadata that was available in that message.

Libsession will do the encryption and return the final payload suitable for sending on the wire. The reverse is available to decrypt an envelope. It supports decrypting both an envelope that is encrypted (e.g.: A groups v2 envelope) and an unencrypted envelope (e.g.: 1o1 and legacy groups envelope) where the content is instead encrypted with the recipient keys.

Still todo:

- Cleaning up the JSON payload generating functions for talking to the Session Pro backend
- A way to sync the revocation list from the Session Pro backend (potentially BYO networking, BYO storage)
- Having a way to transition open groups and open group inbox messages to being able to bundle the pro metadata (they currently send the protobuf `Content` on the wire directly) to enable pro features in those message in a way that is forwards and backwards compatible.
- ~~All the additional helper functions required in the Session Pro PRD (e.g. separating out PRO verification functions so you can call it standalone and evaluate current device PRO status, e.t.c)~~

Doy-lee committed:
Add Session Pro master key derivation

Doy-lee committed:
Add basic pro proof verification functions

Doy-lee committed:
Only give 64b skey from session pro keygen

This simplifies the API and minimises the API surface with code
interacting with it since the public key is already bundled with the secret.

- Add helper function to derive keys from the original seed
- Use _hex_u literals to simplify test construction code

Doy-lee committed:
Apply _hex_u changes to rest of ed25519 tests

Doy-lee committed:
Use cleared_u32 and 64 from sodium_array header

Doy-lee committed:
Linting fix

Doy-lee committed:
Reformat the string split in test_config_userprofile

Doy-lee committed:
Add pro proof and rotating key pair to user profile

Doy-lee committed:
Add c wrappers for Session Pro and proofs

Doy-lee committed:
Update protobufs for Pro and pro proofs

Doy-lee committed:
Typedef array u8x32, add it to types.hpp anad use it in new code

Doy-lee committed:
Add Pro message to Content and decrypt it

Doy-lee committed:
Update Session protobuf files to session-ios at 084e58f

Doy-lee committed:
Revise the PRO protobuf structures

Doy-lee committed:
Port encrypt for namespace into libsession

Doy-lee committed:
Update decrypt function to return the envelope and the metadata

Doy-lee committed:
Add documentation for new session protocol functions

Doy-lee committed:
Add c bindings for new encryption functions

Doy-lee committed:
Revise docs for session protocol c header

Doy-lee committed:
Fix incorrect detection of higher char limits and non-optional pro sigs

Doy-lee committed:
Note that closed groups handles both legacy/non-legacy keys

Doy-lee committed:
Avoid required files in pro messages, handle non-encrypted contents in envelope

Doy-lee committed:
Regen the protobufs w/ non required fields for pro structures

Doy-lee committed:
Allow envelope decrypt to handle encrypted v2 group envelopes

Doy-lee committed:
Do not enforce library types in C interface

Design of the interface should be accomodating not opionated. This also
fixes a bug where the pro backend key wasn't been initialised yet in the
C interface

Doy-lee committed:
Pro version field should use a key that sorts to the top

Due to BT encoding using sort order to order keys. Version at top means
we can conditionally handle version changes easily.

Doy-lee committed:
Fix C session protocol using std::span and missing comments

Doy-lee committed:
Fixup some comments, add missing x25519 pubkey from C decrypted envelope

Doy-lee committed:
Linting

Doy-lee committed:
Fix pro tests breaking due to version change to @

mpretty-cyro commented on include/session/session_protocol.h:
```diff
@@ -0,0 +1,229 @@
+enum {
+    /// Number of characters that a standard message can use. If the message exceeds this then the
+    /// message must activate the higher character limit feature provided by Session Pro which
+    /// allows messages up to 10k characters.
```
Just wondering whether we want to specify that this limit is `utf16` characters?

Doy-lee replied:
Updated and added a footnote that it's currently UTF16 code units but will eventually represent codepoints.

mpretty-cyro commented on include/session/session_protocol.h:
```diff
@@ -0,0 +1,229 @@
+    /// Number of characters that a standard message can use. If the message exceeds this then the
+    /// message must activate the higher character limit feature provided by Session Pro which
+    /// allows messages up to 10k characters.
+    PRO_STANDARD_CHARACTER_LIMIT = 2'000,
```
Should we also have a `PRO_PRO_CHARACTER_LIMIT = 10'000,`? (that way libSession is the source-of-truth for these limits)

There was discussion a little while back on Discord (https://discord.com/channels/408081923835953153/1389855855805206549) about reworking the current name/string length limits and truncation logic to be utf16 based and exposing these limits as constants instead of the current byte limits which I guess would tie into this

Doy-lee replied:
Added a constant for the 10k limit and footnote on the counting mechanism

Bilb commented on include/session/config/namespaces.h:
```diff
@@ -0,0 +1,34 @@
+typedef enum NAMESPACE {
+    // Messages sent to an updated closed group which should be able to be retrieved by revoked
+    // members are stored in this namespace
+    NAMESPACE_REVOKED_RETRIEVABLE_GROUP_MESSAGES = -11,
```
legacy groups have been read only for a while. No platforms should be sending or receiving anything from this namespace anymore, so we should probably just remove it altogether from here, and the `include/session/config/namespaces.hpp` file

mpretty-cyro replied:
I think the namespace we were storing legacy group messages in was `-10` - the `-11` namespace is the one we store kicked messages in for V2 groups

Bilb replied:
oops sorry, all good

mpretty-cyro commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,269 @@
+    /// specified in Destination.
+    ClosedGroup,
+    OpenGroup,
+    OpenGroupInbox,
```
Just wondering whether we want to use `ClosedGroup`/`OpenGroup` terminology here or use `Group`/`Community` terminology which the apps now use? (I guess the former is more accurate to other repos but elsewhere in libSession we've been using `group`/`legacy_group`/`community`)

Doy-lee replied:
Sure, I've updated to group and community, I don't have a distinction of legacy group code in the codebase since that's decided upon by checking the pubkey prefix.

mpretty-cyro commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,269 @@
+enum class EnvelopeType {
+    SessionMessage = ENVELOPE_TYPE_SESSION_MESSAGE,
+    ClosedGroupMessage = ENVELOPE_TYPE_CLOSED_GROUP_MESSGE,
+};
```
We don't actually support sending messages to legacy closed groups anymore on any of the clients (iOS has removed most of the code to handle receiving these messages and I think Android will be doing the same soon) so you could potentially remove `EnvelopeType::ClosedGroupMessage` since it was only used for legacy groups

mpretty-cyro replied:
Actually it looks like we are still setting this for updated groups 🤔 

Confirmed with the Android team - unfortunately they are still using it at the moment so we will need to keep it to avoid breaking old Android versions 🥲

Bilb commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,269 @@
+    /// Both legacy and non-legacy closed groups are to be identified as `ClosedGroup`. A non-legacy
+    /// group is detected by the (0x03) prefix byte on the given `dest_closed_group_pubkey`
+    /// specified in Destination.
+    ClosedGroup,
```
again, legacy groups have been readonly for a while. Sending or receiving a message to/from it should never be done anymore.

Bilb commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,269 @@
+struct Envelope {
+    ENVELOPE_FLAGS flags;
+    EnvelopeType type;
+    std::chrono::milliseconds timestamp;
```
maybe we should make this `timestamp_ms`, just in case :eyes:

Doy-lee replied:
I have the suffix for the C API but C++'s type safety here makes it pretty hard to misuse the timestamp here in a context that isn't compatible with milliseconds.

Bilb commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,269 @@
+    // a Groups v2 envelope or it's re-derived from the Ed25519 pubkey.
+    array_uc32 sender_x25519_pubkey;
+
+    // Status flag for validity of the Session Pro proof embedded in the envelope if it has one.
+    // The status is set to `Nil` if there is no Session Pro proof in the message. Otherwise it's
+    // set to one of the other values to which the remaining pro fields will be populated with data
+    // parsed from the envelope.
+    ProStatus pro_status;
+
+    // The embedded Session Pro proof, only set if the status was not `Nil`.
+    config::ProProof pro_proof;
+
+    // Session Pro bit flag features that were used in the embedded message, only set if the status
+    // was not `Nil`.
+    PRO_FEATURES pro_features;
```
would it make sense to also have `pro_features` unset if the `pro_proof` is not `Valid` ?
It might be the case already, but if yes it would be good to have the comment updated

Doy-lee replied:
Better to preserve the information so that you have the option of using diagnostics on the pro features that were requested but don't have a valid proof anymore. Similarly I don't wipe the pro proof data as I envision basically all client code should be branching on the status before applying pro features.

```cpp
if (pro_status == ProStatus::Valid) {
  // Do sensitive operations w/ pro features,
} else {
  if (pro_features) { /*maybe diagnostics on failure*/ }
}
```

mpretty-cyro commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+    if (!envelope.ParseFromArray(envelope_plaintext.data(), envelope_plaintext.size()))
+        throw std::runtime_error{"Parse envelope from plaintext failed"};
+
+    // Parse type (unconditionallty)
+    if (!envelope.has_type())
+        throw std::runtime_error("Parse envelope failed, missing type");
+
+    switch (envelope.type()) {
+        case SessionProtos::Envelope_Type_SESSION_MESSAGE:
+            result.envelope.type = EnvelopeType::SessionMessage;
+            break;
+
+        case SessionProtos::Envelope_Type_CLOSED_GROUP_MESSAGE:
+            result.envelope.type = EnvelopeType::ClosedGroupMessage;
+            break;
+    }
```
I think Android is the only one doing this now - just wondering whether we should properly deprecate this value and rely on the namespace that the message was received from (if a message with the wrong encryption type is stored in the wrong namespace then it should be considered invalid and dropped anyway)

Doy-lee replied:
For sending we still need to as above for android but for decrypting on receive, you're saying remove this branch and replace it with:
```cpp
if (space == config::Namespace::GroupMessages) {
  result.envelop.type = EnvelopeType::GroupMessage;
} else {
  result.envelope.type = EnvelopeType::SessionMessage;
}
```
Is that right, where the C++ side still wants to keep the envelop type so application code after decryption has all necessary information self-contained in `struct Envelope` to work with, without requiring any external context. But requiring the `type` on the protobuf Envelope is no longer needed (and an outdated practice).

mpretty-cyro replied:
Yea that was my thinking - since the decryption code is based on the `namespace` then we could eventually stop sending the `type` entirely once enough time has passed

> Is that right, where the C++ side still wants to keep the envelop type so application code after decryption has all necessary information self-contained in struct Envelope to work with, without requiring any external context.

I get where this is coming from but since the `Envelope` is encrypted for messages sent to a group the decryption logic is already dependent on knowing the `namespace` that a message came from which makes having a `Envelope.type` feel redundant 😕 (ie. at this point we already know that it's a `GroupMessage`)

Doy-lee replied:
Alright. understood. I have removed the parsing of envelope types on decrypt.

Bilb commented on include/session/types.hpp:
```diff
@@ -1,13 +1,23 @@
+using array_uc64 = std::array<std::uint8_t, 64>;
 
-    using seqno_t = std::int64_t;
+enum class SessionIDPrefix {
+    standard = 0,
+    group = 0x3,
+    community_blinded_legacy = 0x5,
+    community_blinded = 0x15,
+    version_blinded = 0x25,
+    unblinded = 0x7,
+};
```
I am very confused by this. Is it supposed to map to real prefixes?
If yes this should all be changed a fair bit:

```suggestion
enum class SessionIDPrefix {
    standard = 0x05,
    group = 0x03,
    community_unblinded = 0x05,
    community_blinded_legacy = 0x15,
    community_blinded = 0x25,
    version_blinded = 0x07,
};
```

I guess 0x3 is actually 0x03, but the prefixes/names associated are not correct too

mpretty-cyro replied:
The `community_unblinded` uses a `00` prefix which represents the users ed25519 pubkey (and a signature signed with that) whereas the `05` prefix would be for the users x25519 pubkey (and it's signature)

It's used for the SOGS signature header when blinding is disabled - for reference there is a function to get the string prefix for this enum in `internal.hpp`:
```
inline constexpr std::string_view to_string(session::SessionIDPrefix prefix) {
    switch (prefix) {
        case session::SessionIDPrefix::unblinded: return "00"sv;
        case session::SessionIDPrefix::group: return "03"sv;
        case session::SessionIDPrefix::standard: return "05"sv;
        case session::SessionIDPrefix::community_blinded_legacy: return "15"sv;
        case session::SessionIDPrefix::community_blinded: return "25"sv;
        case session::SessionIDPrefix::version_blinded: return "07"sv;
    }

    return "05"sv;  // Fallback to standard, shouldn't occur
};
```

mpretty-cyro commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+                .encrypted = result_internal.encrypted,
+                .ciphertext = result_internal.ciphertext_c,
+        };
+    } catch (...) {
```
Might be nice to expose this error so the iOS client can add errors to it's logs - the convention in a bunch of other places it to pass a `char* error` into the function which, when non-null, is a buffer of 256 bytes

The other option would be to just log the error directly using oxen-logging (as that's already hooked into the logger on iOS)

Doy-lee replied:
I have fixed this, just for the APIs I implement. All the other encryption functions don't do this yet and it's a bit of churn, just updated the parts I've touched for now. I used a more modern C approach to this, these functions take a `char *error, size_t error_len`

Where you pass in your buffer and its length. It doesn't _have_ to be 256 bytes it can be whatever size and it'll write into it and null-terminate it appropriately. It also follows the other C pattern where you can pass in a null buffer and the result will tell you how many bytes are needed to fully receive the error message if there was an error.

For the most part your code can continue to pass in a 256 byte buffer, just that there's no risk writing past the end of the buffer if you accidentally pass in a too small buffer.

mpretty-cyro replied:
Thanks - I think that's fine (it's mostly because these are quite high level functions which could fail for a number of reasons so being able to log the variety of errors that can occur is a bit more important than if a low level function fails, as we can generally log a useful enough error due to the majority of the logic being in the native codce)

Hopefully once iOS 26 comes out we will be able to swap over to the C++ side and don't need to worry about some of this bridging 🤞

mpretty-cyro commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+                sizeof(result.pro_proof.sig));
+
+        result.success = true;
+    } catch (...) {
```
Same as above re. the error

Doy-lee replied:
Updated

Bilb commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+                    enc.mode = Mode::Plaintext;
+                }
+            } else {
+                // Legacy closed groups which have a 05 prefixed key
+                enc.mode = Mode::Envelope;
+                enc.before_envelope_encrypt_for_recipient_deterministic = true;
+                enc.envelope_type =
+                        SessionProtos::Envelope_Type::Envelope_Type_CLOSED_GROUP_MESSAGE;
+                enc.after_envelope = AfterEnvelope::WrapInWSMessage;
+                enc.envelope_src = dest_closed_group_pubkey;
```
to me, this should throw

Doy-lee replied:
I need  confirmation that this branch can't get hit: https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L425

mpretty-cyro replied:
The branch you've linked is the encryption for updated groups, legacy groups would fall into this case: https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L449

The code can be hit by a modified client at the moment, but most of the legacy groups logic has been removed so while a `VisibleMessage` would probably work, anything needed to update a legacy group state would fail - on a non-modified client legacy groups are in a read-only state so there _shouldn't_ be a way to send a message (ie. throwing in this case should be fine)

Doy-lee replied:
Alright, updated, throws if you try and hit that codepath.

Bilb commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+                enc.envelope_type = SessionProtos::Envelope_Type::Envelope_Type_SESSION_MESSAGE;
+                enc.after_envelope = AfterEnvelope::EnvelopeIsCipherText;
+            } else {
+                // Config messages should be sent directly rather than via this method (return just
+                // the plaintext and no-op) See:
+                // https://github.com/session-foundation/session-ios/blob/82deef869d0f7389b799295817f42ad14f8a1316/SessionMessagingKit/Sending%20%26%20Receiving/MessageSender.swift#L498
+                enc.mode = Mode::Plaintext;
+            }
```
would a Config Message not be a DestinationType::SyncMessage instead? :eyes:

Doy-lee replied:
Yes probably an outdated comment I copied from the iOS codebase as that makes sense it'd go into the `SyncMessage` destination. I have removed

mpretty-cyro replied:
Mm yea the iOS codebase doesn't use the standard message sending pipeline for config messages, as a result a config message doesn't have a "destination" the same way other messages do and just gets sent directly to a specific swarm/namespace combination (the "destination" is a container for location information like pubkeys, servers, etc. for non-config messages)

mpretty-cyro commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+}
+
+LIBSESSION_EXPORT
+session_protocol_decrypted_envelope session_protocol_decrypt_envelope(
```
How does the memory work for these functions? (ie. is the caller responsible for freeing it?)

I can see that `span_u8_alloc_or_throw` calls `malloc` which I assume will need to be freed somewhere, and I know there is a bunch of complicated malloc'ing in the codebase to simplify how the memory can be freed by the caller (see `make_string_list` in `internal.hpp` and `contacts_blinded` in `contacts.cpp` - there is another in `network_set_paths_changed_callback` in `session_network.cpp` but I've got a refactored version locally which behaves more similarly to the others)

Doy-lee replied:
Apologies, yes I've documented these allocations now. The spans are allocated with the C runtime's (CRT) `malloc` so it needs to be freed with the equivalent `free` on the `data` member.

mpretty-cyro replied:
Just wondering whether we should add `session_protocol_decrypted_envelope_free`/`session_protocol_encrypted_for_destination_free` functions in case the structures change in the future?

That way the native code just needs to call `result = session_protocol_decrypt_envelope(...)` and `session_protocol_decrypted_envelope_free(result)` and they can trust that the data will be freed correctly, whereas if the native code is responsible for freeing individual variables of the returned object then it'd be pretty easy to miss a change that results in a memory leak

Doy-lee replied:
Yep, update and added.

Bilb commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+    }
+
+    // Parse server timestamp (optional)
+    if (envelope.has_servertimestamp()) {
+        result.envelope.server_timestamp = envelope.servertimestamp();
+        result.envelope.flags |= ENVELOPE_FLAGS_SERVER_TIMESTAMP;
+    }
```
I do not think this is anything used by any platforms,  but to verify.

Bilb commented on src/session_protocol.cpp:
```diff
@@ -0,0 +1,651 @@
+        throw std::runtime_error{fmt::format(
+                "Parse content from envelope failed: {}", result.content_plaintext.size())};
+
+    // A signature must always be present on the envelope. This is to make a pro and non-pro
+    // envelope indistinguishable. If the message does not have pro then this signature must still
+    // be set but will be ignored. So in all instances a signature must be attached (real or
+    // dummy).
+    if (!envelope.has_prosig())
+        throw std::runtime_error("Parse envelope failed, message is missing pro signature");
```
We can't do this, yet. This would throw for any messages already stored in the swarm, and until everyone has upgraded their app, right?

Doy-lee replied:
This is correct, I have fixed it by making it optional now. Added a note that we can enforce it in a year, but for now we can't otherwise it'd break all the old clients as you have mentioned.

Doy-lee committed:
Move group msg encryption primitive into session encrypt, avoids circular dependency

Doy-lee committed:
Move group encryption into session encrypt

This lets the higher-level layer, session protocol to use low level
encryption instead of relying on config to decrypt group messages.

Decryption of a group message is a low-level cryptography routine which
actually belongs in session encrypt.

Doy-lee committed:
Update nomenclamenture, Closed/OpenGroup to Group/Community respectively

Doy-lee committed:
Add some missing typedefs on C structs

Doy-lee committed:
Document the free-eing requirement on the C interface for session protocol/encrypt

Doy-lee committed:
We cannot enforce the presence of a pro signature

We still require it to be optional whilst old clients are sending
envelopes without the signature attached (real or dummy)

jagerman commented on include/session/pro_backend.hpp:
```diff
@@ -0,0 +1,52 @@
+
+namespace session::pro_backend {
+
+constexpr array_uc32 PUBKEY = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                               0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                               0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
```
Appears unused (aside from a static assert, which could simply be changed to `sizeof(array_uc32) == ...` or `array_uc32{}.size() == ...`)

Doy-lee replied:
Updated the static assert and I've added a comment as to what its purpose is. It's meant to be set to the actual Session Pro backend key so that eventually users of the library have the key handy in the interface so as to when they go and verify the proofs they can just pull it from the header.

For libsession I've made the decrypt and verify function parameterise the Session Pro backend public key that it accepts so that you can use it against any arbitrary key (so that it's useful for tests) instead of baking in the key from the header. But the idea is that you can use the key in this header officially for that purpose.

Doy-lee committed:
Don't use assert in test key-gen which breaks in release mode

Doy-lee committed:
Add error logging to C apis

Doy-lee committed:
Add TODO on the purpose of PUBKEY in pro_backend

Doy-lee committed:
Add note that character limit is UTF16 code units for now

Doy-lee committed:
Remove outdated comment, these encrypt dests just receive plaintext

Doy-lee committed:
Remove the CPP session protocol tests, keep C API

The C-API is a wrapper over the C-interface so it thunks into the C++
implementation anyway so it tests both the C and C++ api for us with
a very few exceptions. But for the most part, testing the C API kills
two birds with one stone and reduces the test maintenance overhead of
having to manage 2 test suites in parallel for the same API.

Doy-lee committed:
Fix groups message not encrypting correctly

Doy-lee committed:
Remove envelope type, it is inferred from the namespace

All the payloads in the returned envelope are decrypted so we shouldn't
care too much about the type at that point. The caller can immediately
start working with the contents.

Doy-lee committed:
Encapsulate the freeing of session protocol encrypt/decrypt funcs

Doy-lee committed:
Prohibit the creation of legacy groups messages, no longer supported

Doy-lee committed:
Fix linting

SessionHero01 commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,307 @@
+
+// NOTE: In the CPP file we use C-style enums for bitfields and CPP-style enums for non-bitfield
+// enums where we can to benefit from the type-safety of strong enums.
+//
+// CPP doesn't support named bitfields without casting or operator overloads but C-style
+// enums support it very well. The only issue is that using a native C-style enum enforces some type
+// restrictions that compilers dislike when attempting to manipulate bit fields. For example:
+//
+//   enum Feature {x = 1 << 0, y = 1 << 1}
+//   Feature f = x | y
+//
+// Causes the compiler to complain about trying to do bit ops/assign an unsigned integer to an enum
+// `Feature`. We use a common C pattern/trick by suffixing an underscore to the the original enum,
+// then type define the non-suffixed enum to an unsigned integer:
+//
+//   enum Feature_ {x = 1 << 0, y = 1 << 1}
+//   typedef U64 Feature
+//   Feature f = x | y
+//
+// Does not trigger errors as the underlying type of `f` is actually an unsigned integer. The type
+// define is merely a hint to the user to what flags are to be used when manipulating the variable.
+
+namespace session {
+
+namespace config::groups {
+    class Keys;
+}
+
+enum class ProStatus {
+    // Proof not set
+    Nil = PRO_STATUS_NIL,
+    // Proof set; pro proof sig was not produced by the Pro backend key
+    InvalidProBackendSig = PRO_STATUS_INVALID_PRO_BACKEND_SIG,
+    // Proof set; envelope pro sig was not produced by the Rotating key
+    InvalidUserSig = PRO_STATUS_INVALID_USER_SIG,
+    // Proof set, is verified; has not expired
+    Valid = PRO_STATUS_VALID,
+    // Proof set, is verified; has expired
+    Expired = PRO_STATUS_EXPIRED,
+};
```
Should we leave the expired state out of this status? Whether a pro has expired depends on current time, which I find it the best to leave it to the caller to decide - so that:
1. They can supply their own clock
2. They know when to recheck for validity. 

For this reason I suggest we do these states instead:


```suggestion
enum class ProStatus {
    // Proof not set
    Nil = PRO_STATUS_NIL,
    // Proof set; pro proof sig was not produced by the Pro backend key
    InvalidProBackendSig = PRO_STATUS_INVALID_PRO_BACKEND_SIG,
    // Proof set; envelope pro sig was not produced by the Rotating key
    InvalidUserSig = PRO_STATUS_INVALID_USER_SIG,
    // Proof set, is verified; whether it's expired you should check the expiration time in pro_proof
    Verified = PRO_STATUS_VERIFIED,
};
```

Thoughts?

Doy-lee replied:
Yeah this is currently the case that the caller is in control of the clock.
```cpp
/// - `unix_ts` -- pass in the current system time in seconds which is used to determine, whether or
///   not the Session Pro proof has expired or not if it is in the payload. Ignored if there's no
///   proof in the message.

DecryptedEnvelope decrypt_envelope(
        const DecryptEnvelopeKey& keys,
        std::span<const uint8_t> envelope_payload,
        std::chrono::sys_seconds unix_ts,
        const array_uc32& pro_backend_pubkey);
```
Pro verification in regards to extracting a pro status from the proof is currently embedded in the decryption function but I intend to pull that out to some utility functions some point later so you can call it standalone. I plan to add those after if we're generally happy with the approach here before writing all the helper utilities that go ontop.

Is that sufficient for your use case? Or would you still want to separate the expired-ness from PRO_STATUS.

SessionHero01 replied:
The use case is when `decrypt_envelop` is called, the user hasn't expired yet, let's say 10s later it expires, we won't be calling the `decrypt_envelop` to determine the validity again, we (the application code) will have to arrange its own timer to check the expire timestamp - hence it's best not for the libsession to say whether the pro has expired as it will be tempting to think that as long as the result you get is valid, it's valid forever. Just pushing the expiration check's responsibility to the user forces them to handle it correctly.

This is more about making the API hard to misuse, as I believe at the current form, all the data the application needs is right there. In terms of extracting a pro status from the proof - I think I have some other ideas on the `DecryptedEnvelop` so let's see how you think of that suggestion instead.

SessionHero01 replied:
After a thought I think it doesn't make sense to not have libsession validating the expired status, "expiration" after all, is a criteria of validity. So forget about this one

Doy-lee replied:
Sure, yes I think so we can move expiry out and not merge it into the status field.

Doy-lee replied:
This has been handled out-of-band, we've decided to keep expired in the status, but, reaffirmed that it's useful to have that separated as the clients envision having to check expiry themselves as well independently from decryption at later points.

SessionHero01 commented on include/session/session_protocol.hpp:
```diff
@@ -0,0 +1,307 @@
+    ProStatus pro_status;
+
+    // The embedded Session Pro proof, only set if the status was not `Nil`.
+    config::ProProof pro_proof;
```
Should we have a union type for these instead? So you won't have to worry "only set if status was not Nil":

```cpp
struct Nil {};
struct InvalidSig {
  PRO_FEATURES pro_features_requested;
};

struct InvalidBackendSig : InvalidSig {};
struct InvalidUserSig: InvalidSig {};
struct Valid {
  config::ProProof pro_proof;
  PRO_FEATURES pro_features;
};

typedef std::variant<Nil, InvalidBackendSig, InvalidUserSig, Valid> ProStatus;
```

This `ProStatus` will then be the only property you need on the `DecryptedEnvelop`, and you won't be able to get to any invalid state.

Doy-lee replied:
Handled out-of-band, the solution is to embed the PRO_STATUS into a `std::optional<ProProofWrapper>`-esque structure so that the presence of the optional determines if pro is nil or non-nil.

User code then just needs to branch on the optional to know that there's some, maybe invalid, pro data to work with

```cpp
std::optional<ProProofWrapper> pro_proof_wrapper;
if (pro_proof_wrapper) {
  assert(pro_proof_wrapper->pro_status != NIL); // Cannot be nil by definition because the optional is present. So nil state will be removed from the set of possibilities
}
```

So that removes ProStatus::Nil from the enum and solves the having to worry about a nil status. That is implicit in the presence of the optional

Doy-lee committed:
Remove the need for ProStatus::Nil by using std::optional

Doy-lee committed:
Add docs to pro config

Doy-lee committed:
Add helper functions for getting PRO_STATUS from proof standalone from decryption

Doy-lee committed:
Re-add missing wrapping of 1o1 messages

Was mistakenly removed, upon review of iOS code 1o1s are wrapped in an envelope.

Doy-lee committed:
Mitigate snprintf error prone return value

Doy-lee committed:
Address some outdated doc comments

Doy-lee committed:
Use bytes32/64/33 in session protocol, add wrapper encrypt for {1o1,group,comms} functions

Doy-lee committed:
Remove destinations that don't encrypt

With the new helper functions encrypt_and_wrap_for_{1o1,group,community}
its clearer that this API only really should be called in these
3 circumstances. Making it a kitchen sink black boxes the whole process
but seems to give people options (destinations) that are meaningless as
they don't do anything but we imply they might. I think it's better to
get rid of it and simplify the calling interface.

Doy-lee committed:
Remove namespace from encrypt for destination

We assume that you are encrypting for the destination AND into the
correct namespace. Encrypting for 1o1 as a caller has the intent that
you want to "talk" to someone and in all cases this would have to be
encrypted which is the most obvious outcome of what you would expect to
get if you asked libsession to do it.

The namespace complicated things unnecessarily by making it appear as if
there's an alternative way to encrypt if you were targetting different
namespace. Technically this is true, if you target another
namesspace/mailbox there are different conventions on how an application
should consume data from that mailbox but that is not related to the
task of encrypting for a conversation (1o1, group, community, sync message).

Doy-lee committed:
Linting

Doy-lee committed:
Add missing memory header for zstd

Doy-lee committed:
Move ProProof into session_protocol.hpp

Pro proof is part of the session protocol and is required by some of the
higher-level helper functions in the crypto layer. Because of this we were
having some circular dependency issues, Debian 12 ARM build was complaining that
session_protocol.cpp was calling out to functions in
session/config/pro.h which sits in a layer above us.

This created a circular dependency and upon re-assessment the pro proof
structures really belong in the protocol layer. What does belong in the
config section is the top-level pro config object that contains the
Session Pro rotating key pair and the proof itself.

Doy-lee commented:
Merged as part of https://github.com/session-foundation/libsession-util/pull/64
